Micrometer to OpenTelemetry Bridge

brunobat opened the pull request:
solves https://github.com/quarkusio/quarkus/issues/43621

The implementation uses the [opentelemetry-micrometer-1.5](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/micrometer/micrometer-1.5/library) library.

This allows us to create a Micrometer registry implemented with the OpenTelemetry SDK and APIs.

All telemetry data created with both Micrometer and OpenTelemetry will be processed by the same OpenTelemetry SDK instance from the `quarkus-opentelemetry` extension and will use it's configuration and exporters.

`quarkus-micrometer` extension configurations should also work, as verified on some tests.

Missing:

- [x] Native tests
- [ ] Tests with exemplars using OTel Tracing. 
- [x] Documentation updates
- [x] Decide what to do with the OTel metrics auto instrumentation
- [x] Move the code to the OTel extension? No.
- [x] Disable x.histogram in DistributionSummary
- [x] Silence `io.ope.ins.mic.v1_.OpenTelemetryMeterRegistry] (main) A MeterFilter is being configured ...`
- [x] Enable otel metrics and disable otel metrics instrumentation

brunobat committed:
Micrometer to OTel Bridge

brunobat commented on extensions/micrometer-opentelemetry/deployment/src/main/java/io/quarkus/micrometer/opentelemetry/deployment/MicrometerOtelBridgeProcessor.java:
```diff
@@ -0,0 +1,81 @@
+//        // registered to this registry...
+//        // It's unavoidable because of how Quarkus startup works and users cannot do anything about it.
+//        // see: https://github.com/micrometer-metrics/micrometer/issues/4920#issuecomment-2298348202
+//        systemProperty.produce(
```
@radcortez I wonder why this didn't work here... I had to set these properties in the runtime `application.properties`

brunobat commented:
CC @geoand . This one is ready to review.

geoand commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,219 @@
+- The xref:opentelemetry.adoc[OpenTelemetry Guide] provides information about the OpenTelemetry extension.
+====
+
+The bridge is more than the simple OTLP registry found in Quarkiverse. In this extension, the OpenTelemetry SDK provides a Micrometer registry implementation.
```
I find it odd to start the docs with this phrase. Wouldn't it make sense to start by describing what this extension enabled and then have a note explaining how this is different from the existing Quarkiverse extension?

brunobat replied:
You are right. This start is not user centric. Will improve.

geoand commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,219 @@
+
+The bridge is more than the simple OTLP registry found in Quarkiverse. In this extension, the OpenTelemetry SDK provides a Micrometer registry implementation.
+
+This allows the normal use of the Micrometer API, but have the metrics handled by the OpenTelemetry extension. All the configurations of the OpenTelemetry extension are available for this bridge and enables forwarding to OpenTelemetry all the automatic instrumentation metrics generated by Micrometer in Quarkus, as well as custom user metrics.
```
> This allows the normal use of the Micrometer API, but have the metrics handled by the OpenTelemetry extension

What exactly does this mean? I don't really underastand just by reading this intro

brunobat replied:
Will add an example to illustrate the idea.

geoand commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,219 @@
+
+== Configuration
+
+Micrometer is enabled by default as are OpenTelemetry tracing, metrics and logs.
```
```suggestion
When the extension is present, Micrometer is enabled by default as are OpenTelemetry tracing, metrics and logs.
```

geoand commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,219 @@
+Micrometer is enabled by default as are OpenTelemetry tracing, metrics and logs.
+OpenTelemetry metrics auto-instrumentation for HTTP server and JVM metrics are disabled by default because those metrics are already collected by Micrometer.
+
+Particular automated metrics are not enabled by default and can be enabled by setting the, as an example:
```
```suggestion
Specific automated metrics are not enabled by default and can be enabled by setting the, for example example:
```

geoand replied:
It would also be nice to know exactly what is not enabled

brunobat replied:
All of them. The Micromenter extension requires explicit activation of automated metrics.
Will clarify.

geoand commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,219 @@
+
+When using the `@Timed` annotation with Micrometer, 2 different metrics are https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/324fdbdd452ddffaf2da2c5bf004d8bb3fdfa1dd/instrumentation/micrometer/micrometer-1.5/library/src/main/java/io/opentelemetry/instrumentation/micrometer/v1_5/OpenTelemetryTimer.java#L31[created on the OpenTelemetry side], one `Gauge` for the `max` value ...
+
+The `DistributionSummary` from Micrometer is transformed into a `histogram` and a `DoubleGauge` for the `max` value. If slos are set in the `DistributionSummary` an addition histogram is created for them.
```
`slos`?

brunobat replied:
Will clarify.

geoand commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,219 @@
+
+=== Semantic convention differences
+
+The 2 frameworks follow different semantic conventions and approaches to the concept. The OpenTelemetry metrics are based on the https://opentelemetry.io/docs/concepts/semantic-conventions/[`OpenTelemetry Semantic Conventions`] and are still under active development. Micrometer metrics convention format is around for a long time and has not changed much.
```
> to the concept

What concept?

brunobat replied:
rephrased

geoand commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,219 @@
+
+[NOTE]
+====
+- Some metrics might be absent of the output if they contain no data.
```
```suggestion
- Some metrics might be missing from the output if they contain no data.
```

geoand commented on extensions/micrometer-opentelemetry/deployment/src/main/java/io/quarkus/micrometer/opentelemetry/deployment/MicrometerOTelBridgeConfigBuilderCustomizer.java:
```diff
@@ -0,0 +1,21 @@
+        builder.withSources(
+                new PropertiesConfigSource(Map.of(
+                        "quarkus.otel.metrics.enabled", "true"),
+                        "quarkus-micrometer-opentelemetry", 1),
```
What is `quarkus-micrometer-opentelemetry` and why is the value `1`?

brunobat replied:
It's a mistake. Will fix.

brunobat replied:
Actually, it's right but I will merge it with the next source.
`quarkus-micrometer-opentelemetry` is the config source name and 1 is the priority.

geoand commented on extensions/micrometer-opentelemetry/deployment/src/test/resources/test-logging.properties:
```diff
@@ -0,0 +1,4 @@
+#quarkus.log.category."io.quarkus.micrometer".level=DEBUG
+quarkus.log.category."io.quarkus.bootstrap".level=INFO
+#quarkus.log.category."io.quarkus.arc".level=DEBUG
+quarkus.log.category."io.netty".level=INFO
```
Why these?

brunobat replied:
This was inherited from the original tests in Micrometer.

geoand commented on extensions/micrometer-opentelemetry/runtime/src/main/java/io/quarkus/micrometer/opentelemetry/runtime/MicrometerOtelBridgeRecorder.java:
```diff
@@ -0,0 +1,48 @@
+            @Override
+            public Object apply(SyntheticCreationalContext<Object> context) {
+                if (otelRuntimeConfig.sdkDisabled()) {
+                    return null;
```
Are you sure returning `null` works? cc @mkouba

brunobat replied:
There is a disabled test and it seems to work.

geoand replied:
Very odd. I'd like to hear from @mkouba whether this is expected to work

mkouba replied:
If you ask me whether it's possible to return `null` for a `@Singleton` synthetic bean then I'd say that it could work in theory because it's very similar to a producer method where `IllegalProductException` is only thrown when a non-Dependent (i.e. normal scoped because CDI happily ignores `@Singleton` because of `@ApplicationScoped`) producer returns `null` (that restriction is there for client proxies). 

However, I've never tried that and I think that it's not a good idea in general because then a `null` could be injected. Which is not what most people expect in CDI and thus may result in unexpected NPEs.

It might make more sense to throw an `InjectionException` with a detailed error message. Furthermore, you can instruct the consumers to always inject `Instance<BEAN>` and check `Instance#isResolvable()` before use.

CC @manovotn @Ladicek

geoand replied:
Thanks @mkouba! 

> However, I've never tried that and I think that it's not a good idea in general because then a null could be injected.

My thoughts exactly!

> Furthermore, you can instruct the consumers to always inject Instance<BEAN> and check Instance#isResolvable() before use.

Right, I don't think returning `null` is ever necessary

brunobat replied:
@mkouba the point is that I want to return nothing if the sdk is disabled, I don't want to abort with an exception... I imagine I'll need to do that in the processor.

mkouba replied:
> the point is that I want to return nothing if the sdk is disabled

But then the consumers of your bean would need to expect `null` injected... :shrug:

brunobat replied:
Only the OTel SDK would expect this... and it's disabled. That's why it will not blow up on null.

geoand commented on extensions/micrometer-opentelemetry/runtime/src/main/java/io/quarkus/micrometer/opentelemetry/runtime/MicrometerOtelBridgeRecorder.java:
```diff
@@ -0,0 +1,48 @@
+@Recorder
+public class MicrometerOtelBridgeRecorder {
+
+    public Function<SyntheticCreationalContext<Object>, Object> createBridge(OTelRuntimeConfig otelRuntimeConfig) {
```
Let's not use `Object` here as we know the types

brunobat replied:
right.

geoand reviewed (commented):
Nice work!

I've added some questions and suggestions

geoand commented:
The CI failure seems related

radcortez commented on extensions/micrometer-opentelemetry/deployment/src/main/java/io/quarkus/micrometer/opentelemetry/deployment/MicrometerOTelBridgeConfigBuilderCustomizer.java:
```diff
@@ -0,0 +1,21 @@
+                new PropertiesConfigSource(Map.of(
+                        "quarkus.otel.metrics.enabled", "true"),
+                        "quarkus-micrometer-opentelemetry", 1),
+                new PropertiesConfigSource(Map.of(
```
Could both sources be merged in a single `PropertiesConfigSource`?

brunobat replied:
yes, will fix this.

brunobat commented:
Already addresses the pending reviews.

geoand commented on extensions/micrometer-opentelemetry/deployment/src/test/java/io/quarkus/micrometer/opentelemetry/deployment/compatibility/JvmCompatibilityTest.java:
```diff
@@ -0,0 +1,105 @@
+                                .hasAttributesSatisfying(attributes -> attributes.isEmpty())));
+    }
+
+    //    @Test
+    //    void printAll() {
+    //        metricExporter.assertCountDataPointsAtLeastOrEqual("jvm.threads.started", null, 1);
+    //        final List<MetricData> metricDataList = metricExporter.getFinishedMetricItems();
+    //
+    //        metricDataList.forEach(System.out::println);
+    //    }
```
I assume this should be removed?

brunobat replied:
yes

geoand commented on extensions/micrometer-opentelemetry/runtime/src/main/resources/application.properties:
```diff
@@ -0,0 +1,3 @@
+# Tweak logs:
+quarkus.log.category."io.opentelemetry.instrumentation.micrometer.v1_5.OpenTelemetryMeterRegistry".level=ERROR
+quarkus.log.category."io.micrometer.core.instrument.composite.CompositeMeterRegistry".level=ERROR
```
Why are we doing this? In any case, if an extension is going to be changing log levels for categories, it should be done via `LogCategoryBuildItem`

brunobat replied:
This is due to a warning that micrometer throws when we configure a filter too late.
I tried many things to suppress those messages and this was the only thing that worked. I tried something like this:

```
    @BuildStep
    public void tuneDefaultConfigs(BuildProducer<RunTimeConfigurationDefaultBuildItem> runtimeConfigProducer,
            BuildProducer<LogCategoryBuildItem> logCategoryProducer) {

        // Suppress noisy logs from Micrometer:
        // ...A MeterFilter is being configured after a Meter has been registered to this registry...
        logCategoryProducer.produce(new LogCategoryBuildItem(
                "io.opentelemetry.instrumentation.micrometer.v1_5.OpenTelemetryMeterRegistry",
                Level.SEVERE,
                true));
        logCategoryProducer.produce(new LogCategoryBuildItem(
                "io.micrometer.core.instrument.composite.CompositeMeterRegistry",
                Level.SEVERE,
                true));
    }
```

And I get the messages:
```
2025-01-21 09:25:21,201 WARN  [io.mic.cor.ins.com.CompositeMeterRegistry] (main) A MeterFilter is being configured after a Meter has been registered to this registry. All MeterFilters should be configured before any Meters are registered. If that is not possible or you have a use case where it should be allowed, let the Micrometer maintainers know at https://github.com/micrometer-metrics/micrometer/issues/4920. Enable DEBUG level logging on this logger to see a stack trace of the call configuring this MeterFilter.
```

geoand replied:
Did you try:

```java
new LogCategoryBuildItem(
                "io.opentelemetry.instrumentation.micrometer.v1_5.OpenTelemetryMeterRegistry",
                org.jboss.logmanager.Level.ERROR))
```

?

brunobat replied:
Tried it now. It didn't work. 
Maybe the log message is triggered too soon.

geoand replied:
I need to check this out. How do I reproduce the problem?

brunobat replied:
On the PR use this method in `io.quarkus.micrometer.opentelemetry.deployment.MicrometerOtelBridgeProcessor`

```
    @BuildStep
    public void tuneDefaultConfigs(BuildProducer<RunTimeConfigurationDefaultBuildItem> runtimeConfigProducer,
            BuildProducer<LogCategoryBuildItem> logCategoryProducer) {
        runtimeConfigProducer.produce(
                new RunTimeConfigurationDefaultBuildItem("quarkus.otel.instrument.http-server-metrics", "false"));
        runtimeConfigProducer.produce(
                new RunTimeConfigurationDefaultBuildItem("quarkus.otel.instrument.jvm-metrics", "false"));

        // Suppress noisy logs from Micrometer:
        // ...A MeterFilter is being configured after a Meter has been registered to this registry...
        logCategoryProducer.produce(new LogCategoryBuildItem(
                "io.opentelemetry.instrumentation.micrometer.v1_5.OpenTelemetryMeterRegistry",
                org.jboss.logmanager.Level.ERROR,
                true));
        logCategoryProducer.produce(new LogCategoryBuildItem(
                "io.micrometer.core.instrument.composite.CompositeMeterRegistry",
                org.jboss.logmanager.Level.ERROR,
                true));
    }
```
Comment out the `application.properties` contents from runtime and run this IT test:

`io.quarkus.micrometer.opentelemetry.MicrometerCounterInterceptorTest`

geoand replied:
PR updated with a more appropriate fix

geoand commented on integration-tests/micrometer-opentelemetry/pom.xml:
```diff
@@ -0,0 +1,203 @@
+            <artifactId>assertj-core</artifactId>
+            <scope>compile</scope>
+        </dependency>
+<!--        <dependency>-->
+<!--            <groupId>io.quarkus</groupId>-->
+<!--            <artifactId>quarkus-hibernate-orm-panache-deployment</artifactId>-->
+<!--            <version>${project.version}</version>-->
+<!--            <type>pom</type>-->
+<!--            <scope>test</scope>-->
+<!--            <exclusions>-->
+<!--                <exclusion>-->
+<!--                    <groupId>*</groupId>-->
+<!--                    <artifactId>*</artifactId>-->
+<!--                </exclusion>-->
+<!--            </exclusions>-->
+<!--        </dependency>-->
+<!--        <dependency>-->
+<!--            <groupId>io.quarkus</groupId>-->
+<!--            <artifactId>quarkus-jdbc-h2-deployment</artifactId>-->
+<!--            <version>${project.version}</version>-->
+<!--            <type>pom</type>-->
+<!--            <scope>test</scope>-->
+<!--            <exclusions>-->
+<!--                <exclusion>-->
+<!--                    <groupId>*</groupId>-->
+<!--                    <artifactId>*</artifactId>-->
+<!--                </exclusion>-->
+<!--            </exclusions>-->
+<!--        </dependency>-->
```
Should this be removed?

geoand committed:
Ensure logging is setup before Micrometer configures registries

This is done because we need sometimes need to tweak the
logging categories and the corresponding build items
have no effect if logging has not yet been setup

brunobat commented:
Thanks @geoand :)

geoand commented:
🙏🏽

brunobat commented:
rebased

brunobat commented:
There's something wrong with the branch... It's not piking up the conflicts on `native-tests.json`.
Do not merge.

brunobat commented:
All good after all. Holly's native-tests.json reorg is not on main and I got confused.
@gsmet are you ok with this one being merged?

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
````diff
@@ -0,0 +1,233 @@
+The extension allows the normal use of the Micrometer API, but have the metrics handled by the OpenTelemetry extension.
+
+As an example, the `@Timed` annotation from Micrometer is used to measure the execution time of a method:
+```java
````
Please use proper `[source]` blocks. There are more to adjust below.

brunobat replied:
done

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
````diff
@@ -0,0 +1,233 @@
+```
+quarkus.micrometer.binder.jvm=true
+```
+on the `application.properties` file.
````
```suggestion
in the `application.properties` file.
```

brunobat replied:
done

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+
+For this and other properties you can use with the extension, Please refer to:
+
+* xref:telemetry-micrometer.adoc#configuration-reference[Micrometer metrics configurations]
```
```suggestion
* xref:telemetry-micrometer.adoc#configuration-reference[Micrometer metrics configuration properties]
```

brunobat replied:
done

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+For this and other properties you can use with the extension, Please refer to:
+
+* xref:telemetry-micrometer.adoc#configuration-reference[Micrometer metrics configurations]
+* xref:opentelemetry.adoc#configuration-reference[OpenTelemetry configurations]
```
```suggestion
* xref:opentelemetry.adoc#configuration-reference[OpenTelemetry configuration properties]
```

brunobat replied:
done

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+
+When using the `@Timed` annotation with Micrometer, 2 different metrics are https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/324fdbdd452ddffaf2da2c5bf004d8bb3fdfa1dd/instrumentation/micrometer/micrometer-1.5/library/src/main/java/io/opentelemetry/instrumentation/micrometer/v1_5/OpenTelemetryTimer.java#L31[created on the OpenTelemetry side], one `Gauge` for the `max` value ...
+
+The `DistributionSummary` from Micrometer is transformed into a `histogram` and a `DoubleGauge` for the `max` value. If service level objectives (slo) are set to `true` when creating a `DistributionSummary`, an additional histogram is created for them.
```
Not entirely sure but seems consistent with above:

```suggestion
The `DistributionSummary` from Micrometer is transformed into a `Histogram` and a `DoubleGauge` for the `max` value. If service level objectives (slo) are set to `true` when creating a `DistributionSummary`, an additional histogram is created for them.
```

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+
+=== Semantic convention differences
+
+The 2 frameworks follow different semantic conventions. The OpenTelemetry Metrics are based on the https://opentelemetry.io/docs/concepts/semantic-conventions/[`OpenTelemetry Semantic Conventions`] and are still under active development (early 2025). Micrometer metrics convention format is around for a long time and has not changed much.
```
Let's avoid using backticks for things that don't really need them.

```suggestion
The 2 frameworks follow different semantic conventions. The OpenTelemetry Metrics are based on the https://opentelemetry.io/docs/concepts/semantic-conventions/[OpenTelemetry Semantic Conventions] and are still under active development (early 2025). Micrometer metrics convention format is around for a long time and has not changed much.
```

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
````diff
@@ -0,0 +1,233 @@
+```
+The JVM and HTTP server metrics are collected by Micrometer.
+
+Next, are examples of the metrics collected by Micrometer and a comparison of what would be the `quarkus-micrometer-registry-prometheus` output vs the one on this bridge. A link to the equivalent OpenTelemetry Semantic Convention is also provided for reference and is not currently used in the bridge.
````
```suggestion
Next, are examples of the metrics collected by Micrometer and a comparison of what would be the `quarkus-micrometer-registry-prometheus` output vs the one of this bridge. A link to the equivalent OpenTelemetry Semantic Convention is also provided for reference and is not currently used in the bridge.
```

brunobat replied:
I'll keep my version if you don't mind.

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+
+[NOTE]
+====
+- Some metrics might be missing from the output if they contain no data.
```
```suggestion
Some metrics might be missing from the output if they contain no data.
```

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+== See the output
+
+=== Grafana-OTel-LGTM Dev Service
+You can use the xref:observability-devservices-lgtm.adoc[Grafana-OTel-LGTM] devservice.
```
```suggestion
You can use the xref:observability-devservices-lgtm.adoc[Grafana-OTel-LGTM] Dev Service.
```

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+=== Grafana-OTel-LGTM Dev Service
+You can use the xref:observability-devservices-lgtm.adoc[Grafana-OTel-LGTM] devservice.
+
+This Dev service includes a Grafana for visualizing data, Loki to store logs, Tempo to store traces and Prometheus to store metrics.
```
```suggestion
This Dev Service includes Grafana for visualizing data, Loki to store logs, Tempo to store traces and Prometheus to store metrics.
```

gsmet commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+You can use the xref:observability-devservices-lgtm.adoc[Grafana-OTel-LGTM] devservice.
+
+This Dev service includes a Grafana for visualizing data, Loki to store logs, Tempo to store traces and Prometheus to store metrics.
+Also provides and OTel collector to receive the data.
```
```suggestion
It also provides an OTel collector to receive the data.
```

gsmet commented on extensions/micrometer-opentelemetry/deployment/pom.xml:
```diff
@@ -0,0 +1,134 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-micrometer-opentelemetry</artifactId>
+            <version>${project.version}</version>
```
Drop the version as it's the best way to make sure it's properly in the BOM.

gsmet commented on integration-tests/micrometer-opentelemetry/pom.xml:
```diff
@@ -0,0 +1,177 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-micrometer-opentelemetry</artifactId>
+            <version>999-SNAPSHOT</version>
```
Again please drop the version. It also generates noise when releasing as it needs updating.

gsmet commented on integration-tests/micrometer-opentelemetry/pom.xml:
```diff
@@ -0,0 +1,177 @@
+            <artifactId>quarkus-rest-jackson</artifactId>
+        </dependency>
+
+        <!-- Verify that the client can co-exist with the server -->
```
Does it make sense in the context of this test or was it copied from another test?

brunobat replied:
It does, because it uses a rest client.

gsmet commented on integration-tests/micrometer-opentelemetry/pom.xml:
```diff
@@ -0,0 +1,177 @@
+            <scope>test</scope>
+        </dependency>
+        <!-- Needed for InMemorySpanExporter to verify captured traces -->
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-sdk-testing</artifactId>
+        </dependency>
```
Is it a test dependency? If so make sure the scope is explicit.

brunobat replied:
It is explicit because it bundles an in-memory exporter needed at the main runtime.
This is on purpose.

gsmet commented on integration-tests/micrometer-opentelemetry/pom.xml:
```diff
@@ -0,0 +1,177 @@
+                        <configuration>
+                            <skipTests>${native.surefire.skip}</skipTests>
+                            <systemPropertyVariables>
+                                <vertx.disableWebsockets>false</vertx.disableWebsockets>
```
Why do we need this?

brunobat replied:
Clear copy paste error. Will fix.

gsmet reviewed (commented):
I added a few minor suggestions, questions, mostly to the doc.

I would recommend fixing all the easy stuff before merging and then we can discuss the remaining points after merge if they require some discussion.

brunobat commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+:topics: observability,opentelemetry,metrics,micrometer,tracing,logs
+:extensions: io.quarkus:quarkus-micrometer-opentelemetry
+
+This extension provides support to both `Micrometer` and `OpenTelemetry` in Quarkus applications. It streamlines integration by incorporating both extensions along with a bridge that enables sending Micrometer metrics via OpenTelemetry.
```
done

brunobat commented on docs/src/main/asciidoc/telemetry-micrometer-to-opentelemetry.adoc:
```diff
@@ -0,0 +1,233 @@
+
+This reduces the overhead of having an independent Micrometer registry plus the OpenTelemetry SDK in memory for the same application when both `quarkus-micrometer` and `quarkus-opentelemetry` extensions are used independently.
+
+*The OpenTelemetry SDK will handle all metrics.* Either Micrometer metrics (manual or automatic) and OpenTelemetry Metrics can be used . All are available with this single extension.
```
done

brunobat commented:
Addressed the review @gsmet

gsmet commented:
I just saw this had the backport label, I don't think we can backport it, it's too much of a change, it's going to be for 3.19.

brunobat commented:
That's fine.
